### ISSUE TITLE
volume services broken if nsync not available on cc_bridge

### ISSUE THREAD

**Issue by silvestre** [cloudfoundry/cloud_controller_ng#892]
## Issue

In our environment volume services stop working if the nsync job is not available on Diego's cc_bridge.

## Context

When starting an application how the cloud controller communicates with Diego is decided in
https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/diego/messenger.rb#L40-L46
```ruby        
if bypass_bridge?
          app_recipe_builder = AppRecipeBuilder.new(config: config, process: process)
          DesireAppHandler.create_or_update_app(process_guid, app_recipe_builder, bbs_apps_client)
        else
          desire_message = protocol.desire_app_request(process, config[:default_health_check_timeout])
          nsync_client.desire_app(process_guid, desire_message)
end
```

If the nsync job is not available on the cc_bridge, direct communication with the BBS is done through the bbs_apps_client. When nsync is available, the nsync_client is used.

When using the bbs_apps_client (not nsync) the Desire LRP request has the following volume_services structure:

```JSON                  
 "volume_mounts": [
                        {
                            "container_dir": "/var/vcap/data/d0c4396f-b2cb-4224-b5c3-10dbc5dc2539",
                            "mode": "rw",
                            "shared": {
                                "volume_id": "d0c4396f-b2cb-4224-b5c3-10dbc5dc2539",
                                "mount_config": "..."
                            }
                        }
                    ],
```

Here the driver field is missing and on BBS side the request fails validation:
https://github.com/cloudfoundry/bbs/blob/7acac8a24d7a1d07167ef0bf63af05f7426d4e88/models/volume_mount.go#L46-L48

## Steps to Reproduce

1. Do not deploy nsync
2. Bind an application to a volume service
3. Start the application

## Expected result

The application should start

## Current result

The application does not start. Errors are not reported to the end user but inside a cloud_controller api VM you can find the following log entry:

```JSON
{"timestamp":1504281007.145622,"message":"Request failed: 500: {\"description\"=>\"Runner invalid request: invalid volume_mount driver\", \"error_code\"=>\"CF-RunnerInvalidRequest\", \"code\"=>170014, \"test_mode_info\"=>{\"description\"=>\"Runner invalid request: invalid volume_mount driver\", \"error_code\"=>\"CF-RunnerInvalidRequest\", \"backtrace\"=>[\"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/bbs_apps_client.rb:122:in `runner_invalid_request!'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/bbs_apps_client.rb:16:in `block in desire_app'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/bbs_apps_client.rb:107:in `handle_diego_errors'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/bbs_apps_client.rb:12:in `desire_app'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/desire_app_handler.rb:11:in `create_or_update_app'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/messenger.rb:44:in `send_desire_request'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/runner.rb:21:in `block in start'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/runner.rb:42:in `with_logging'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/diego/runner.rb:21:in `start'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/app_observer.rb:38:in `react_to_state_change'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/app_observer.rb:23:in `block in updated'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/app_observer.rb:46:in `with_diego_communication_handling'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/app_observer.rb:21:in `updated'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/models/runtime/process_model.rb:516:in `block in after_save'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:401:in `each'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:401:in `remove_transaction'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:248:in `_transaction'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:199:in `block in transaction'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:165:in `transaction'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/v2/app_update.rb:13:in `update'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/app/controllers/runtime/apps_controller.rb:257:in `update'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/app/controllers/base/base_controller.rb:79:in `dispatch'\", \"/var/vcap/data/packages/cloud_controller_ng/bfc0c08e0d7e505a2a75483f6f54bdecaab3c1de/cloud_controller_ng/lib/cloud_controller/rest_controller/routes.rb:16:in `block in define_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1610:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1610:in `block in compile!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:975:in `block (3 levels) in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:994:in `route_eval'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:975:in `block (2 levels) in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1015:in `block in process_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1013:in `catch'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1013:in `process_route'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:973:in `block in route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:972:in `each'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:972:in `route!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1085:in `block in dispatch!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `block in invoke'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `catch'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `invoke'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1082:in `dispatch!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:907:in `block in call!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `block in invoke'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `catch'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:1067:in `invoke'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:907:in `call!'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:895:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb:18:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/path_traversal.rb:16:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/json_csrf.rb:18:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-protection-1.5.3/lib/rack/protection/frame_options.rb:31:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/nulllogger.rb:9:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/head.rb:13:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:182:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/sinatra-1.4.7/lib/sinatra/base.rb:2013:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/urlmap.rb:66:in `block in call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/urlmap.rb:50:in `each'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/urlmap.rb:50:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/rate_limiter.rb:43:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb:22:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/security_context_setter.rb:19:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/vcap_request_id.rb:15:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/cors.rb:49:in `call_app'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/cors.rb:14:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_metrics.rb:12:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/rack-1.6.8/lib/rack/builder.rb:153:in `call'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/thin-1.7.0/lib/thin/connection.rb:86:in `block in pre_process'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/thin-1.7.0/lib/thin/connection.rb:84:in `catch'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/thin-1.7.0/lib/thin/connection.rb:84:in `pre_process'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/thin-1.7.0/lib/thin/connection.rb:50:in `block in process'\", \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.3.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:1067:in `block in spawn_threadpool'\"]}}","log_level":"error","source":"cc.api","data":{"request_guid":"4b665dce-264d-4e1f-5856-c16223309dfb::93e5b49d-eb4e-4250-9438-d7870158b8d7"},"thread_id":47041350137000,"fiber_id":47041381627540,"process_id":9300,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/lib/sinatra/vcap.rb","lineno":53,"method":"block in registered"}
```

## Possible Fix

Not a fix, but a workaround is to deploy nsync.

**Comment by matt-royal** [comment]
@silvestre we were able to reproduce your error when the `temporary_local_*` fields are `true` and that the issue does not appear when set to `false`. We'll start looking into a fix now.

**Comment by zrob** [comment]
fix is pushed in fd814d90cb0256be21b0949eeb34d3aa669f2b07

will be available in capi release 1.42.0
